src/twisted/web/http.py

```diff
@@ -1386,7 +1386,7 @@ def getClientIP(self):
         @returns: the client IP address
         @rtype: C{str}
         """
-        if isinstance(self.client, address.IPv4Address):
+        if isinstance(self.client, (address.IPv4Address, address.IPv6Address)):
             return self.client.host
         else:
             return None
```

Comment:
I know that this is existing code, but when is expected to return `None`.

The docstring is not clear.

Can this ever return `None`? In which case? Is there a test for this?

The code coverate, reports that this branch is executed  but I could not found the explicit tests https://codecov.io/gh/twisted/twisted/src/89e71172c2e4b2f923d3f3fa49922d50262394df/src/twisted/web/http.py#L1392

I saw that it is called in twisted.web.test.test_web.test_etagMatched due to a log which is trigerred

Reply:
> Can this ever return None?

It absolutely can!  Try this:

```twist web --port=unix:/tmp/sock --wsgi wsgiref.simple_server.demo_app```

Connect to it like so:

```curl --unix-socket /tmp/sock http://example.invalid/```

Now you'll get a traceback from ```twisted.web```:

```
...
          File "/home/mrw/src/python/twisted/src/twisted/web/wsgi.py", line 303, in __init__
            'REMOTE_ADDR': _wsgiString(request.getClientIP()),
          File "/home/mrw/src/python/twisted/src/twisted/web/wsgi.py", line 85, in _wsgiString
            return string.decode("iso-8859-1")
        builtins.AttributeError: 'NoneType' object has no attribute 'decode'
```

This is mentioned in [a comment on 7704](https://twistedmatrix.com/trac/ticket/7704#comment:2), which was linked to from [this ticket](https://twistedmatrix.com/trac/ticket/8241) and also captured in [5406](https://twistedmatrix.com/trac/ticket/5406).